src/gfn/samplers.py

```diff
@@ -77,36 +441,47 @@ def sample_actions(
             - Optional tensor of log probabilities (if save_logprobs=True)
             - Optional tensor of estimator outputs (if save_estimator_outputs=True)
         """
-        # TODO: Should estimators instead ignore None for the conditioning vector?
-        if conditioning is not None:
-            with has_conditioning_exception_handler("estimator", self.estimator):
-                estimator_output = self.estimator(states, conditioning)
-        else:
-            with no_conditioning_exception_handler("estimator", self.estimator):
-                estimator_output = self.estimator(states)
+        if ctx is None:
+            ctx = self.adapter.init_context(
+                batch_size=states.batch_shape[0],
+                device=states.device,
+                conditioning=conditioning,
+            )
 
-        dist = self.estimator.to_probability_distribution(
-            states, estimator_output, **policy_kwargs
+        step_mask = torch.ones(
+            states.batch_shape[0], dtype=torch.bool, device=states.device
         )
+        dist, ctx = self.adapter.compute(states, ctx, step_mask, **policy_kwargs)
 
         with torch.no_grad():
-            actions = dist.sample()
+            actions_tensor = dist.sample()
 
         if save_logprobs:
-            log_probs = dist.log_prob(actions)
+            log_probs = dist.log_prob(actions_tensor)
             if torch.any(torch.isinf(log_probs)):
                 raise RuntimeError("Log probabilities are inf. This should not happen.")
         else:
             log_probs = None
 
-        actions = env.actions_from_tensor(actions)
+        # Allow adapter to record per-step artifacts for callers that reuse ctx.
+        self.adapter.record_step(
+            ctx=ctx,
+            step_mask=step_mask,
+            sampled_actions=actions_tensor,
+            dist=dist,
+            save_logprobs=save_logprobs,
+            save_estimator_outputs=save_estimator_outputs,
+        )
```

Comment:
why can't be done every `.compute` step?

Reply:
they could maybe be merged, but I was thinking it was nice to have them in two different places (perhaps you're happy with the default record step logic but not with the default compute logic). the idea is that on average you shouldn't have to write too much code.